### PR TITLE
XSD and Schematron Validation

### PR DESCRIPTION
**Pull request by sjpadgett**
- create new validation class
- fix bug in existing Schematron class.

This continues to implement document validation based upon various schemas depending on document type CCDA, QRDA etc.
I still have reporting to user and logging.

### CODE REVIEW

**Comment by sjpadgett** [comment]
Example of test CCD. The errors reported should be ignore but this schema errors on nullFlavors when it is acceptable. Unknown how I'll handle.

```
Array
(
    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.4.16-2014-06-09-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:substanceAdministration[cda:templateId[@root='2.16.840.1.113883.10.20.22.4.16' and @extension='2014-06-09']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => @moodCode and @moodCode=document('voc.xml')/voc:systems/voc:system[@valueSetOid='2.16.840.1.113883.11.20.9.18']/voc:code/@value
                                            [message] => SHALL contain exactly one [1..1] @moodCode, which SHALL be selected from ValueSet MoodCodeEvnInt urn:oid:2.16.840.1.113883.11.20.9.18 STATIC 2011-04-03 (CONF:1098-7497).
                                            [path] => /*/*[28]/*/*[2]/*/*[6]/*
                                        )

                                    [1] => stdClass Object
                                        (
                                            [test] => cda:effectiveTime[@xsi:type='IVL_TS']
                                            [message] => SHALL contain exactly one [1..1] effectiveTime (CONF:1098-7508) such that it
                                            [path] => /*/*[28]/*/*[2]/*/*[6]/*
                                        )

                                )

                        )

                )

        )

    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.4.7-2014-06-09-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:observation[cda:templateId[@root='2.16.840.1.113883.10.20.22.4.7' and @extension='2014-06-09']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => cda:effectiveTime[count(cda:low)=1]
                                            [message] => This effectiveTime SHALL contain exactly one [1..1] low (CONF:1098-31538).
                                            [path] => /*/*[28]/*/*[1]/*/*[6]/*/*[7]/*
                                        )

                                )

                        )

                )

        )

    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.4.40-2014-06-09-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:encounter[cda:templateId[@root='2.16.840.1.113883.10.20.22.4.40' and @extension='2014-06-09']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => @moodCode and @moodCode=document('voc.xml')/voc:systems/voc:system[@valueSetOid='2.16.840.1.113883.11.20.9.23']/voc:code/@value
                                            [message] => SHALL contain exactly one [1..1] @moodCode, which SHALL be selected from ValueSet Planned moodCode (Act/Encounter/Procedure) urn:oid:2.16.840.1.113883.11.20.9.23 STATIC 2014-09-01 (CONF:1098-8565).
                                            [path] => /*/*[28]/*/*[13]/*/*[6]/*
                                        )

                                )

                        )

                )

        )

    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.4.52-2015-08-01-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:substanceAdministration[cda:templateId[@root='2.16.840.1.113883.10.20.22.4.52' and @extension='2015-08-01']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => @moodCode and @moodCode=document('voc.xml')/voc:systems/voc:system[@valueSetOid='2.16.840.1.113883.11.20.9.18']/voc:code/@value
                                            [message] => SHALL contain exactly one [1..1] @moodCode, which SHALL be selected from ValueSet MoodCodeEvnInt urn:oid:2.16.840.1.113883.11.20.9.18 STATIC (CONF:1198-8827).
                                            [path] => /*/*[28]/*/*[5]/*/*[6]/*
                                        )

                                )

                        )

                )

        )

    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.1.1-2015-08-01-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:ClinicalDocument[cda:templateId[@root='2.16.840.1.113883.10.20.22.1.1' and @extension='2015-08-01']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => not(cda:documentationOf/cda:serviceEvent/cda:performer) or cda:documentationOf/cda:serviceEvent/cda:performer[@typeCode and @typeCode=document('voc.xml')/voc:systems/voc:system[@valueSetOid='2.16.840.1.113883.1.11.19601']/voc:code/@value]
                                            [message] => The performer, if present, SHALL contain exactly one [1..1] @typeCode, which SHALL be selected from ValueSet x_ServiceEventPerformer urn:oid:2.16.840.1.113883.1.11.19601 STATIC (CONF:1198-14840).
                                            [path] => /*
                                        )

                                )

                        )

                )

        )

    [#p-urn-hl7ii-2.16.840.1.113883.10.20.22.4.30-2015-08-01-errors] => stdClass Object
        (
            [title] => 
            [rules] => Array
                (
                    [0] => stdClass Object
                        (
                            [context] => cda:act[cda:templateId[@root='2.16.840.1.113883.10.20.22.4.30' and @extension='2015-08-01']]
                            [errors] => Array
                                (
                                    [0] => stdClass Object
                                        (
                                            [test] => cda:statusCode[@code and @code=document('voc.xml')/voc:systems/voc:system[@valueSetOid='2.16.840.1.113883.11.20.9.19']/voc:code/@value]
                                            [message] => This statusCode SHALL contain exactly one [1..1] @code, which SHALL be selected from ValueSet ProblemAct statusCode urn:oid:2.16.840.1.113883.11.20.9.19 STATIC (CONF:1198-19086).
                                            [path] => /*/*[28]/*/*[1]/*/*[6]/*
                                        )

                                )

                        )

                )

        )

    [#p-urn-oid-2.16.840.1.113883.10.20.22.5.2-warnings] => stdClass Object ...
       

)

```

File: interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CarecoordinationTable.php [hunk 1]
```diff
@@ -22,6 +22,7 @@
 use Laminas\Db\TableGateway\AbstractTableGateway;
 use OpenEMR\Services\Cda\CdaTemplateImportDispose;
 use OpenEMR\Services\Cda\CdaTemplateParse;
+use OpenEMR\Services\Cda\CdaValidateDocuments;
```
Comment (sjpadgett) [inline interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CarecoordinationTable.php:25]: @adunsulag Note I've made changes if you're in this class

**Comment by sjpadgett** [comment]
This is nuts so, now I find this project! I may revert to it.
https://github.com/ewadkins/cda-schematron

File: interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CarecoordinationTable.php (hunk 1 above)
Comment (adunsulag) [reply interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Model/CarecoordinationTable.php:25]: I haven't touched that class, but thanks for giving me the heads up.

**Comment by sjpadgett** [comment]
Can not get schema document(voc.xml) to reference the rule OID for vocabulary items.
Try everything for file pathing. May also explain some of the issues with report xml references to its stylesheet.

Have I mentioned recently just how much I hate xml(x-tremely m-essedup l-anguage):)

**Comment by bradymiller** [comment]
what other services do we have in node modules? (note the main directory node_modules is removed during production build process since just using it to collect js packages and build themes)

**Comment by sjpadgett** [comment]
all our cqm services reside in node modules. what happens to all our npm installed packages?

**Comment by bradymiller** [comment]
they are brought over to public/assets/
I was under impression that the cqm services (except for ccda service which is in it's own place) are for development purposes (ie. rebuilding the cqm template and mustache stuff). Is there anything that are needed for production? If so, then actually may be better to have the live services in ccdaservices node modules

**Comment by sjpadgett** [comment]
nope the oe-sqm-service is for running our calculator. you're thinking of the oe-module-cqm that is used to gen patient models for new reporting periods. currently the projects get copied to public but their node dependencies, not!

there is no doubt we're going to need an install package to run on install. there are to many node dependencies to put in public IMO. you may have other idea I'm not seeing.

I may be able to install the schematron along with ccda service but still, all of this needs to get streamlined...

**Pull request closed without merging**